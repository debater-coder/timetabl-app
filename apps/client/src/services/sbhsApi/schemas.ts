import { DateTime } from "luxon";
import { z } from "zod";

/**
 * Possible SBHS API endpoints
 */
export type SbhsApiEndpoint =
  | "calendar/days.json"
  | "calendar/terms.json"
  | "timetable/bells.json"
  | "barcodenews/list.json"
  | "dailynews/list.json" // Used in Announcements
  | "diarycalendar/events.json"
  | "details/particiaption.json"
  | "details/userinfo.json" // Used in Barcodes
  | "timetable/daytimetable.json" // Used in Home
  | "timetable/timetable.json";

export const profileSchema = z.object({
  username: z.coerce.string(),
  studentId: z.coerce.string(),
  givenName: z.coerce.string(),
  surname: z.coerce.string(),
  rollClass: z.coerce.string(),
  yearGroup: z.coerce.string(),
  role: z.coerce.string(), // may be valid for staff
  department: z.coerce.string(), // may be valid for staff
  office: z.coerce.string(), // may be valid for staff
  email: z.coerce.string().email(),
  decEmail: z.coerce.string().email(),
});

export enum NoticeYear {
  ALL,
  YEAR7,
  YEAR8,
  YEAR9,
  YEAR10,
  YEAR11,
  YEAR12,
  STAFF,
  UNKNOWN,
}

export const yearMapping: Record<string, NoticeYear> = {
  "7": NoticeYear.YEAR7,
  "8": NoticeYear.YEAR8,
  "9": NoticeYear.YEAR9,
  "10": NoticeYear.YEAR10,
  "11": NoticeYear.YEAR11,
  "12": NoticeYear.YEAR12,
  staff: NoticeYear.STAFF,
  Staff: NoticeYear.STAFF,
};

export const noticesSchema = z
  .object({
    date: z.coerce.number(), // UNIX timestamp
    dayInfo: z.object({
      date: z.coerce.string(), // YYYY-MM-DD
      term: z.coerce.string(), // the school term
      week: z.coerce.string().nullish(), // the week of term
      weekType: z.coerce.string().nullish(), // the week type
      dayNumber: z.coerce.string().nullish(), // day in timetable cycle
    }),
    dateYesterday: z.coerce.number(), // previous school day. UNIX timestamp
    dateTomorrow: z.coerce.number(), // next school day. UNIX timestamp
    notices: z.array(
      z.object({
        // array of notices
        title: z.coerce.string(), // title of the daily notice
        content: z.coerce.string(), // body of the notice (HTML encoded)
        years: z.array(z.coerce.string()), // applicable year groups for the notice
        dates: z.array(z.coerce.string()).nullish(), // array of dates (YYYY-MM-DD) notice will appear
        relativeWeight: z.coerce.number().nullish(), // a priority indicator for the notice
        isMeeting: z.coerce
          .number()
          .transform((isMeeting) => !!isMeeting)
          .nullish(), // Boolean. Meetings have intrinsic +1 relative weight
        meetingDate: z.coerce.string().nullish(), // date of meeting (YYYY-MM-DD), null if not a meeting
        meetingTimeParsed: z.coerce.string().nullish(), // if parsable meeting time converted to HH:MM:SS
        meetingTime: z.coerce.string().nullish(), // "time" the user set the meeting for (string)
        displayYears: z.coerce.string().nullish(), // a nice way of showing message applicability
        authorName: z.coerce.string(), // display name of user who created message
        date: z.coerce.string().optional(), // date notice was created (YYYY-MM-DD) (specific to Timetabl News Only)
      })
    ),
  })
  .transform((notices) => {
    return notices.notices.map((notice) => ({
      ...notice,
      years: notice.years.map(
        (year) => yearMapping?.[year] ?? NoticeYear.UNKNOWN
      ),
    }));
  });

export type TimetablNotice = z.output<typeof noticesSchema>[0];

export const bellSchema = z.object({
  time: z.string(),
  startTime: z.string(),
  endTime: z.string().nullish(),
  bell: z.string(),
  bellDisplay: z.string(),
  period: z.string(),
});

export const subjectSchema = z.object({
  colour: z.string(),
  fullTeacher: z.string(),
  subject: z.string(),
  title: z.string(),
});

export const periodSchema = z.object({
  fullTeacher: z.string().nullish(),
  teacher: z.string().nullish(),
  title: z.string().nullish(),
  year: z.string().nullish(),
  room: z.string().nullish(),
});

const formatCasual = (casual?: string | null) => {
  if (!casual) return null;
  if (casual.length === 0) return casual;
  if (casual.length === 1) return `${casual.toUpperCase()}.`;

  return `${casual[
    casual.length - 1
  ]?.toUpperCase()} ${casual[0]?.toUpperCase()}${casual
    .substring(1, casual.length - 1)
    .toLowerCase()}.`;
};

export type TimetablPeriod = {
  name: string;
  shortName: string;
  endTime: string;
  startTime: string;
  date: string;
  room?: string;
  teacher?: string;
  colour?: string;
  key: string;
  casual?: string;
  roomTo?: string;
};

export type TimetablDtt = {
  periods: TimetablPeriod[];
  date: string;
};

export const dttSchema = z
  .object({
    bells: z.array(bellSchema).nullable(),
    timetable: z.union([
      z.object({
        subjects: z.record(subjectSchema),
        timetable: z.object({
          periods: z.record(periodSchema),
        }),
      }),
      z.literal(false),
    ]),
    date: z.string(),
    classVariations: z.union([
      z.record(
        z.object({
          period: z.string().nullish(),
          year: z.string().nullish(),
          title: z.string().nullish(),
          teacher: z.string().nullish(),
          type: z.enum(["nocover", "replacement", "novariation"]),
          casual: z.string().nullish(),
          casualSurname: z.string().nullish(),
        })
      ),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      z.array(z.undefined()).max(0),
    ]),
    roomVariations: z.union([
      z.record(
        z.object({
          roomTo: z.string(),
        })
      ),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      z.array(z.undefined()).max(0),
    ]),
  })
  .transform((data): TimetablDtt => {
    const classVariations = data.classVariations;
    const roomVariations = data.roomVariations;

    const result = {
      periods: (data.bells ?? [])
        .flatMap((bell, index, bells) => {
          const timetable = data.timetable;

          if (timetable === false) {
            throw Error("This situation should never happen");
          }

          const subjects = timetable.subjects;
          const period = timetable.timetable.periods?.[bell?.bell];

          let subject;
          let casual;
          let roomTo;

          let shortName = bell.bell;
          let name = bell?.bellDisplay;
          const teacher = period?.fullTeacher ?? period?.teacher ?? undefined;

          if (period?.title) {
            name = period?.title;

            if (period?.year) {
              name = period?.year + name;
              shortName = name;
              subject = subjects?.[name] ?? subject;

              name = subject?.title ?? name;
            }
          }

          if (
            !Array.isArray(classVariations) &&
            classVariations?.[bell?.period] &&
            classVariations?.[bell?.period]?.type !== "novariation"
          ) {
            casual =
              classVariations?.[bell?.period]?.casualSurname ??
              formatCasual(classVariations?.[bell?.period]?.casual) ??
              "No one";
          }

          if (
            !Array.isArray(roomVariations) &&
            roomVariations?.[bell?.period]
          ) {
            roomTo = roomVariations?.[bell?.period]?.roomTo ?? "-";
          }

          return [
            {
              name: "Transition to " + name,
              endTime: DateTime.fromISO(
                `${data?.date}T${bell?.startTime}`
              ).toISO(),
              startTime: DateTime.fromISO(
                bells?.[index - 1]?.endTime ?? "00:00"
              ).toISO(),
              date: data?.date,
              key: `${bell.bell}-transition`,
              shortName: "transition",
            },
            {
              name,
              room: period?.room ?? undefined,
              teacher,
              startTime: DateTime.fromISO(
                `${data?.date}T${bell?.startTime}`
              ).toISO(),
              endTime: DateTime.fromISO(
                `${data?.date}T${bell?.endTime}`
              ).toISO(),
              colour:
                subject?.colour && period?.room
                  ? `#${subject?.colour}`
                  : "transparent",
              key: bell.bell,
              casual,
              roomTo,
              date: data?.date,
              shortName,
            },
          ];
        })
        .filter((period) => period?.startTime !== period?.endTime),
      date: DateTime.fromISO(data?.date).toISO(),
    };

    return result;
  });

export const sbhsKey =
  (endpoint: SbhsApiEndpoint) =>
  <T extends Record<string, string>>(options?: T) =>
    [`/sbhs/${endpoint}`, options] as const;

export const daySchema = z.record(
  z.object({
    date: z.coerce.string(),
    term: z.coerce.string(),
    week: z.coerce.string(),
    weekType: z.coerce.string(),
    dayNumber: z.coerce.string(),
  })
);
