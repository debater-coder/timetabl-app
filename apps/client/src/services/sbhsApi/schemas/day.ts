import { z } from "zod";

export const daySchema = z.record(
  z.object({
    date: z.string(),
    term: z.coerce.number().nullish(),
    week: z.coerce.number().nullish(),
    weekType: z.string().length(1).nullish(),
    dayNumber: z.coerce.number().nullish(),
  })
);
