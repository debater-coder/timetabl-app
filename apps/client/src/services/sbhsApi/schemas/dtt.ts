import { DateTime } from "luxon";
import { z } from "zod";

export const bellSchema = z.object({
  time: z.string(),
  startTime: z.string(),
  endTime: z.string().nullish(),
  bell: z.string(),
  bellDisplay: z.string(),
  period: z.string(),
});

export const subjectSchema = z.object({
  colour: z.string(),
  fullTeacher: z.string(),
  subject: z.string(),
  title: z.string(),
});

export const periodSchema = z.object({
  fullTeacher: z.string().nullish(),
  teacher: z.string().nullish(),
  title: z.string().nullish(),
  year: z.string().nullish(),
  room: z.string().nullish(),
});

const formatCasual = (casual?: string | null) => {
  if (!casual) return null;
  if (casual.length === 0) return casual;
  if (casual.length === 1) return `${casual.toUpperCase()}.`;

  return `${casual[
    casual.length - 1
  ]?.toUpperCase()} ${casual[0]?.toUpperCase()}${casual
    .substring(1, casual.length - 1)
    .toLowerCase()}.`;
};

export type TimetablPeriod = {
  name: string;
  shortName: string;
  endTime: string;
  startTime: string;
  date: string;
  room?: string;
  teacher?: string;
  colour?: string;
  key: string;
  casual?: string;
  roomTo?: string;
};

export type TimetablDtt = {
  periods: TimetablPeriod[];
  date: string;
};

export const dttSchema = z
  .object({
    bells: z.array(bellSchema).nullable(),
    timetable: z.union([
      z.object({
        subjects: z.record(subjectSchema),
        timetable: z.object({
          periods: z.record(periodSchema),
        }),
      }),
      z.literal(false),
    ]),
    date: z.string(),
    classVariations: z.union([
      z.record(
        z.object({
          period: z.string().nullish(),
          year: z.string().nullish(),
          title: z.string().nullish(),
          teacher: z.string().nullish(),
          type: z.enum(["nocover", "replacement", "novariation"]),
          casual: z.string().nullish(),
          casualSurname: z.string().nullish(),
        })
      ),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      z.array(z.undefined()).max(0),
    ]),
    roomVariations: z.union([
      z.record(
        z.object({
          roomTo: z.string(),
        })
      ),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      z.array(z.undefined()).max(0),
    ]),
  })
  .transform((data): TimetablDtt => {
    const classVariations = data.classVariations;
    const roomVariations = data.roomVariations;

    return {
      periods: (data.bells ?? [])
        .flatMap((bell, index, bells) => {
          const timetable = data.timetable;

          if (timetable === false) {
            throw Error("This situation should never happen");
          }

          const subjects = timetable.subjects;
          const period = timetable.timetable.periods?.[bell?.bell];

          let subject;
          let casual;
          let roomTo;

          let shortName = bell.bell;
          let name = bell?.bellDisplay;
          const teacher = period?.fullTeacher ?? period?.teacher ?? undefined;

          if (period?.title) {
            name = period?.title;

            if (period?.year) {
              name = period?.year + name;
              shortName = name;
              subject = subjects?.[name] ?? subject;

              name = subject?.title ?? name;
            }
          }

          if (
            !Array.isArray(classVariations) &&
            classVariations?.[bell?.period] &&
            classVariations?.[bell?.period]?.type !== "novariation"
          ) {
            casual =
              classVariations?.[bell?.period]?.casualSurname ??
              formatCasual(classVariations?.[bell?.period]?.casual) ??
              "No one";
          }

          if (
            !Array.isArray(roomVariations) &&
            roomVariations?.[bell?.period]
          ) {
            roomTo = roomVariations?.[bell?.period]?.roomTo ?? "-";
          }

          return [
            {
              name: "Transition to " + name,
              endTime: DateTime.fromISO(
                `${data?.date}T${bell?.startTime}`
              ).toISO(),
              startTime: DateTime.fromISO(
                bells?.[index - 1]?.endTime ?? "00:00"
              ).toISO(),
              date: data?.date,
              key: `${bell.bell}-transition`,
              shortName: "transition",
            },
            {
              name,
              room: period?.room ?? undefined,
              teacher,
              startTime: DateTime.fromISO(
                `${data?.date}T${bell?.startTime}`
              ).toISO(),
              endTime: DateTime.fromISO(
                `${data?.date}T${bell?.endTime}`
              ).toISO(),
              colour:
                subject?.colour && period?.room
                  ? `#${subject?.colour}`
                  : "transparent",
              key: bell.bell,
              casual,
              roomTo,
              date: data?.date,
              shortName,
            },
          ];
        })
        .filter((period) => period?.startTime !== period?.endTime),
      date: DateTime.fromISO(data?.date).toISO(),
    };
  });
