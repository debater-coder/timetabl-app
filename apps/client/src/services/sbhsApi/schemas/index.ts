/**
 * Possible SBHS API endpoints
 */
export type SbhsApiEndpoint =
  | "calendar/days.json"
  | "calendar/terms.json"
  | "timetable/bells.json"
  | "barcodenews/list.json"
  | "dailynews/list.json" // Used in Announcements
  | "diarycalendar/events.json"
  | "details/particiaption.json"
  | "details/userinfo.json" // Used in Barcodes
  | "timetable/daytimetable.json" // Used in Home
  | "timetable/timetable.json";

export const sbhsKey =
  (endpoint: SbhsApiEndpoint) =>
  <T extends Record<string, string>>(options?: T) =>
    [`/sbhs/${endpoint}`, options] as const;
