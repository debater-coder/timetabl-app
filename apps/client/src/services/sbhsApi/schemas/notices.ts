import { z } from "zod";

export enum NoticeYear {
  ALL,
  YEAR7,
  YEAR8,
  YEAR9,
  YEAR10,
  YEAR11,
  YEAR12,
  STAFF,
  UNKNOWN,
}

export const yearMapping: Record<string, NoticeYear> = {
  "7": NoticeYear.YEAR7,
  "8": NoticeYear.YEAR8,
  "9": NoticeYear.YEAR9,
  "10": NoticeYear.YEAR10,
  "11": NoticeYear.YEAR11,
  "12": NoticeYear.YEAR12,
  staff: NoticeYear.STAFF,
  Staff: NoticeYear.STAFF,
};

export const noticesSchema = z
  .object({
    date: z.coerce.number(), // UNIX timestamp
    dayInfo: z.object({
      date: z.coerce.string(), // YYYY-MM-DD
      term: z.coerce.string(), // the school term
      week: z.coerce.string().nullish(), // the week of term
      weekType: z.coerce.string().nullish(), // the week type
      dayNumber: z.coerce.string().nullish(), // day in timetable cycle
    }),
    dateYesterday: z.coerce.number(), // previous school day. UNIX timestamp
    dateTomorrow: z.coerce.number(), // next school day. UNIX timestamp
    notices: z.array(
      z.object({
        // array of notices
        title: z.coerce.string(), // title of the daily notice
        content: z.coerce.string(), // body of the notice (HTML encoded)
        years: z.array(z.coerce.string()), // applicable year groups for the notice
        dates: z.array(z.coerce.string()).nullish(), // array of dates (YYYY-MM-DD) notice will appear
        relativeWeight: z.coerce.number().nullish(), // a priority indicator for the notice
        isMeeting: z.coerce
          .number()
          .transform((isMeeting) => !!isMeeting)
          .nullish(), // Boolean. Meetings have intrinsic +1 relative weight
        meetingDate: z.coerce.string().nullish(), // date of meeting (YYYY-MM-DD), null if not a meeting
        meetingTimeParsed: z.coerce.string().nullish(), // if parsable meeting time converted to HH:MM:SS
        meetingTime: z.coerce.string().nullish(), // "time" the user set the meeting for (string)
        displayYears: z.coerce.string().nullish(), // a nice way of showing message applicability
        authorName: z.coerce.string(), // display name of user who created message
        date: z.coerce.string().optional(), // date notice was created (YYYY-MM-DD) (specific to Timetabl News Only)
      })
    ),
  })
  .transform((notices) => {
    return notices.notices.map((notice) => ({
      ...notice,
      years: notice.years.map(
        (year) => yearMapping?.[year] ?? NoticeYear.UNKNOWN
      ),
    }));
  });

export type TimetablNotice = z.output<typeof noticesSchema>[0];
