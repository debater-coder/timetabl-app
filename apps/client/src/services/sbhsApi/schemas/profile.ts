import { z } from "zod";

export const profileSchema = z.object({
  username: z.coerce.string(),
  studentId: z.coerce.string(),
  givenName: z.coerce.string(),
  surname: z.coerce.string(),
  rollClass: z.coerce.string(),
  yearGroup: z.coerce.string(),
  role: z.coerce.string(), // may be valid for staff
  department: z.coerce.string(), // may be valid for staff
  office: z.coerce.string(), // may be valid for staff
  email: z.coerce.string().email(),
  decEmail: z.coerce.string().email(),
});
