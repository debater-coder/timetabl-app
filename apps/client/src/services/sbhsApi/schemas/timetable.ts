import { z } from "zod";

const timetableDaySchema = z.object({
  dayname: z.string(), // name of the day
  routine: z.string(), // routine (see timetable/daytimetable()
  rollcall: z.object({
    // roll call info for this day
    title: z.string(), // name of roll class
    teacher: z.string().nullish(), // teacher code for roll class
    room: z.string().nullish(), // room for roll call
  }),
  periods: z.record(
    z.object({
      // corresponds to routine
      title: z.string(), // short name for class
      teacher: z.string().nullish(), // teacher code for class
      room: z.string().nullish(), // room for class
      year: z.string().nullish(), // year for class (Note [1])
    })
  ),
});

export type TimetableDay = z.infer<typeof timetableDaySchema>;
export const timetableSchema = z
  .object({
    student: z.object({
      surname: z.string(),
      givenname: z.string(),
      sex: z.string().nullish(), // "M"/"B" for male
      DOB: z.string().nullish(), // date of birth. Unix timestamp
      roll: z.coerce.string().nullish(), // index of roll class
      lines: z.record(z.coerce.number()).nullish(),
      extraLines: z.record(z.coerce.number()).nullish(),
      BoSNumber: z.coerce.number(), // Board of Studies number. 0 if not available
      studentId: z.string(), // Student ID number
      year: z.string(), // student's [primary] year group
      years: z.string().array(), // array of years the student is in
    }),
    days: z.record(timetableDaySchema),
    subjects: z.array(
      z.object({
        title: z.string(), // title of the class
        shortTitle: z.string(), // corresponds to period.title
        subject: z.string(), // full name of the subject
        teacher: z.string().nullish(), // teacher code for the class teacher
        fullTeacher: z.string().nullish(), // full name of the teacher
        year: z.coerce.string().nullish(), // year for the class (Note [1])
      })
    ),
    extraSubjects: z
      .record(
        z.object({
          // these subjects have no timetable classes
          title: z.string(), // title of the class
          shortTitle: z.string(), // corresponds to period.title
          teacher: z.string().nullish(), // teacher code for the class teacher
          fullTeacher: z.string().nullish(), // full name of the teacher
        })
      )
      .nullish(),

    rollcall: z.object({
      // roll class details
      title: z.string(), // title of the class
      teacher: z.string(), // teacher code for the teacher
      fullTeacher: z.string(), // full name of the teacher
      room: z.string().nullish(), // roll class room
    }),
    advisor: z.string(), // year adviser
  })
  .transform((data) => {
    return Object.values(data.days);
  });
