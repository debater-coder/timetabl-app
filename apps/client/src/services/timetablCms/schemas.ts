import { z } from "zod";
import { NoticeYear } from "../sbhsApi/schemas";

export const timetablNewsSchema = z
  .object({
    data: z.array(
      z.object({
        attributes: z.object({
          title: z.string(),
          content: z.string(),
          createdAt: z.string(),
          createdBy: z.object({
            data: z.object({
              attributes: z.object({
                firstname: z.string(),
                lastname: z.string(),
              }),
            }),
          }),
        }),
      })
    ),
  })
  .transform((data) => {
    return data.data.map(
      (announcement: {
        attributes: {
          title: string;
          content: string;
          createdAt: string;
          createdBy: {
            data: {
              attributes: {
                firstname: string;
                lastname: string;
              };
            };
          };
        };
      }) => ({
        ...announcement.attributes,
        years: [
          NoticeYear.YEAR7,
          NoticeYear.YEAR8,
          NoticeYear.YEAR9,
          NoticeYear.YEAR10,
          NoticeYear.YEAR11,
          NoticeYear.YEAR12,
          NoticeYear.STAFF,
        ],
        authorName:
          announcement.attributes.createdBy.data.attributes.firstname +
          " " +
          announcement.attributes.createdBy.data.attributes.lastname,
        date: announcement.attributes.createdAt,
      })
    );
  });
