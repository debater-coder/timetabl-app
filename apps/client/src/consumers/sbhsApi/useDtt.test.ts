import { dttSchema } from "./schemas";
import { describe, expect, test } from "vitest";

describe("the schema", () => {
  test("parses a day timetable with no periods", () => {
    const dtt = {
      status: "OK",
      date: "2023-09-20",
      bells: [
        {
          period: "0",
          startTime: "08:00",
          endTime: "08:57",
          type: "O",
          time: "08:00",
          bell: "0",
          bellDisplay: "Period 0",
        },
        {
          period: "RC",
          startTime: "08:57",
          endTime: "09:00",
          type: "O",
          time: "08:57",
          bell: "RC",
          bellDisplay: "Roll Call",
        },
        {
          period: "1",
          startTime: "09:00",
          endTime: "10:00",
          type: "T",
          time: "09:00",
          bell: "1",
          bellDisplay: "Period 1",
        },
        {
          period: "2",
          startTime: "10:05",
          endTime: "11:05",
          type: "T",
          time: "10:05",
          bell: "2",
          bellDisplay: "Period 2",
        },
        {
          period: "R",
          startTime: "11:05",
          endTime: "11:22",
          type: "R",
          time: "11:05",
          bell: "R",
          bellDisplay: "Recess",
        },
        {
          period: "3",
          startTime: "11:25",
          endTime: "12:25",
          type: "T",
          time: "11:25",
          bell: "3",
          bellDisplay: "Period 3",
        },
        {
          period: "WFL1",
          startTime: "12:25",
          endTime: "12:45",
          type: "R",
          time: "12:25",
          bell: "WFL1",
          bellDisplay: "Lunch 1",
        },
        {
          period: "WFL2",
          startTime: "12:45",
          endTime: "13:02",
          type: "R",
          time: "12:45",
          bell: "WFL2",
          bellDisplay: "Lunch 2",
        },
        {
          period: "4",
          startTime: "13:05",
          endTime: "14:05",
          type: "T",
          time: "13:05",
          bell: "4",
          bellDisplay: "Period 4",
        },
        {
          period: "5",
          startTime: "14:10",
          endTime: "15:10",
          type: "T",
          time: "14:10",
          bell: "5",
          bellDisplay: "Period 5",
        },
        {
          period: "EoD",
          startTime: "15:10",
          endTime: null,
          type: "O",
          time: "15:10",
          bell: "EoD",
          bellDisplay: "End of Day",
        },
      ],
      timetable: {
        timetable: {
          periods: [],
        },
        subjects: [],
      },
      roomVariations: [],
      classVariations: [],
      serverTimezone: "36000",
      shouldDisplayVariations: true,
    };

    expect(dttSchema.parse(dtt)).toMatchSnapshot();
  });
});
