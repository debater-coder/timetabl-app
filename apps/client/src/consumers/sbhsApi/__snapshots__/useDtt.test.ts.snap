// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`the schema > parses a day timetable with no periods 1`] = `
{
  "date": "2023-09-20T00:00:00.000+10:00",
  "periods": [
    {
      "date": "2023-09-20",
      "endTime": "2023-09-20T08:00:00.000+10:00",
      "key": "0-transition",
      "name": "Transition to Period 0",
      "shortName": "transition",
      "startTime": "2023-09-20T00:00:00.000+10:00",
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T08:57:00.000+10:00",
      "key": "0",
      "name": "Period 0",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "0",
      "startTime": "2023-09-20T08:00:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T09:00:00.000+10:00",
      "key": "RC",
      "name": "Roll Call",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "RC",
      "startTime": "2023-09-20T08:57:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T10:00:00.000+10:00",
      "key": "1",
      "name": "Period 1",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "1",
      "startTime": "2023-09-20T09:00:00.000+10:00",
      "teacher": undefined,
    },
    {
      "date": "2023-09-20",
      "endTime": "2023-09-20T10:05:00.000+10:00",
      "key": "2-transition",
      "name": "Transition to Period 2",
      "shortName": "transition",
      "startTime": "2023-09-20T10:00:00.000+10:00",
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T11:05:00.000+10:00",
      "key": "2",
      "name": "Period 2",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "2",
      "startTime": "2023-09-20T10:05:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T11:22:00.000+10:00",
      "key": "R",
      "name": "Recess",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "R",
      "startTime": "2023-09-20T11:05:00.000+10:00",
      "teacher": undefined,
    },
    {
      "date": "2023-09-20",
      "endTime": "2023-09-20T11:25:00.000+10:00",
      "key": "3-transition",
      "name": "Transition to Period 3",
      "shortName": "transition",
      "startTime": "2023-09-20T11:22:00.000+10:00",
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T12:25:00.000+10:00",
      "key": "3",
      "name": "Period 3",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "3",
      "startTime": "2023-09-20T11:25:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T12:45:00.000+10:00",
      "key": "WFL1",
      "name": "Lunch 1",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "WFL1",
      "startTime": "2023-09-20T12:25:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T13:02:00.000+10:00",
      "key": "WFL2",
      "name": "Lunch 2",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "WFL2",
      "startTime": "2023-09-20T12:45:00.000+10:00",
      "teacher": undefined,
    },
    {
      "date": "2023-09-20",
      "endTime": "2023-09-20T13:05:00.000+10:00",
      "key": "4-transition",
      "name": "Transition to Period 4",
      "shortName": "transition",
      "startTime": "2023-09-20T13:02:00.000+10:00",
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T14:05:00.000+10:00",
      "key": "4",
      "name": "Period 4",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "4",
      "startTime": "2023-09-20T13:05:00.000+10:00",
      "teacher": undefined,
    },
    {
      "date": "2023-09-20",
      "endTime": "2023-09-20T14:10:00.000+10:00",
      "key": "5-transition",
      "name": "Transition to Period 5",
      "shortName": "transition",
      "startTime": "2023-09-20T14:05:00.000+10:00",
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": "2023-09-20T15:10:00.000+10:00",
      "key": "5",
      "name": "Period 5",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "5",
      "startTime": "2023-09-20T14:10:00.000+10:00",
      "teacher": undefined,
    },
    {
      "casual": undefined,
      "colour": "transparent",
      "date": "2023-09-20",
      "endTime": null,
      "key": "EoD",
      "name": "End of Day",
      "room": undefined,
      "roomTo": undefined,
      "shortName": "EoD",
      "startTime": "2023-09-20T15:10:00.000+10:00",
      "teacher": undefined,
    },
  ],
}
`;